blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitScm.java

```diff
@@ -25,21 +34,97 @@
 import net.sf.json.JSONObject;
 import org.eclipse.jgit.lib.Repository;
 import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.Stapler;
+import org.kohsuke.stapler.StaplerRequest;
 import org.kohsuke.stapler.json.JsonBody;
 
 import javax.annotation.Nonnull;
 import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.util.List;
+import java.util.Locale;
 
 public class GitScm extends AbstractScm {
+
     public static final String ID = "git";
 
+    static final String CREDENTIAL_DOMAIN_NAME = "blueocean-git-domain";
+    static final String CREDENTIAL_DESCRIPTION_PW = "Git username/password";
+
     protected final Reachable parent;
 
     public GitScm(Reachable parent) {
         this.parent = parent;
     }
 
+    /**
+     * Create the credentialId for a specific repositoryUrl (which will be normalized)
+     * @param repositoryUrl
+     * @return credentialId string
+     */
+    public static String makeCredentialId(String repositoryUrl) {
+
+        final String normalizedUrl = normalizeServerUrl(repositoryUrl);
+
+        try {
+            final java.net.URI uri = new URI(normalizedUrl);
+
+            // Require a host
+            String host = uri.getHost();
+            if (host == null || host.length() == 0) {
+                return null;
+            }
+
+            // Only http(s) urls have a default credential ID keyed to the repo right now
+            String scheme = uri.getScheme();
+            if (scheme != null && scheme.startsWith("http")) {
+                return ID + ":" + normalizedUrl;
```

Comment:
This is problematic as the credential id is used in classic credential management to manage the credential. e.g. Click on User->Credentials and then click on created credential id, it redirects you to `http://localhost:8080/jenkins/user/admin/credentials/store/user/domain/blueocean-git-domain/credential/git:https://github.com/vivek/pipeline-demo-test.git`. At this point stapler craps out as it tries to find route to github resulting on 404. As alternative, a hash of the computed id can work?

![image](https://user-images.githubusercontent.com/38139/39275187-4086207a-4899-11e8-8838-e98715f40ae7.png)

Reply:
Seems like a bug in classic not urlEncoding it. Easy enough to use a hash, but if we change the ID to gibberish, how will the user know which "bob/****" credential is the right one when he wants to delete it?

Reply:
> Seems like a bug in classic not urlEncoding it. 

Yeah, I think its bug in credentials-plugin that does not encode the id. Anyways, user can get to it via clicking credential domain so probably not a show stopper. I am little concerned as changing credential id has backward compatibility concerns. That is in future if we change it, then we have to either migrate existing job configurations to include new changed id or somehow make system work with old id. Basically whether we use hash or current id implementation, we should make sure we think bit more about it. cc: @stephenc 

> Easy enough to use a hash, but if we change the ID to gibberish, how will the user know which "bob/****" credential is the right one when he wants to delete it?

Right. I played it little bit more. Maybe we can use description for that. The domain page shows description, so choosing a hash should probably be fine.

![image](https://user-images.githubusercontent.com/38139/39280422-3aadc070-48b4-11e8-9584-942bb4ba662c.png)

Reply:
Ack

Reply:
@sophistifunk You are using repo URL, e.g. `git:https://github.com/vivek/pipeline-demo-test.git`. This means for each repo you are going to create a new credential? Should it not be just the host:port? I wonder if normalized id could be `git:https:github.com`. This can avoid hashing too. but may be there is use case where we need to create Id of repo URL?